corehq/ex-submodules/couchforms/dbaccessors.py

```diff
@@ -0,0 +1,97 @@
+from couchforms.models import XFormInstance, doc_types
+from django.conf import settings
+
+
+def get_form_ids_by_type(domain, type_, start=None, end=None):
+    assert type_ in doc_types()
+    startkey = [domain, 'by_type', type_]
+    if end:
+        endkey = startkey + end.isoformat()
+    else:
+        endkey = startkey + [{}]
+
+    if start:
+        startkey.append(start.isoformat())
+
+    return [row['id'] for row in XFormInstance.get_db().view(
+        "couchforms/all_submissions_by_domain",
+        startkey=startkey,
+        endkey=endkey,
+        reduce=False,
+    )]
+
+
+def get_forms_by_type(domain, type_, recent_first=False,
+                      limit=None):
+    assert type_ in doc_types()
+    # no production code should be pulling all forms in one go!
+    assert limit is not None
+    startkey = [domain, 'by_type', type_]
+    endkey = startkey + [{}]
+    if recent_first:
+        startkey, endkey = endkey, startkey
+    return XFormInstance.view(
+        "couchforms/all_submissions_by_domain",
+        startkey=startkey,
+        endkey=endkey,
+        reduce=False,
+        descending=recent_first,
+        include_docs=True,
+        limit=limit,
+        classes=doc_types(),
+    ).all()
+
+
+def get_forms_of_all_types(domain):
+    assert settings.UNIT_TESTING
+    startkey = [domain, 'by_type']
+    endkey = startkey + [{}]
+    return XFormInstance.view(
+        "couchforms/all_submissions_by_domain",
+        startkey=startkey,
+        endkey=endkey,
+        reduce=False,
+        include_docs=True,
+        classes=doc_types(),
+    ).all()
+
+
+def get_number_of_forms_by_type(domain, type_):
+    assert type_ in doc_types()
+    startkey = [domain, 'by_type', type_]
+    endkey = startkey + [{}]
+
+    return XFormInstance.view(
+        "couchforms/all_submissions_by_domain",
+        startkey=startkey,
+        endkey=endkey,
+        reduce=True,
+    ).one()['value']
+
+
+def get_number_of_forms_of_all_types(domain):
+    startkey = [domain, 'by_type']
+    endkey = startkey + [{}]
+
+    return XFormInstance.view(
+        "couchforms/all_submissions_by_domain",
+        startkey=startkey,
+        endkey=endkey,
+        reduce=True,
+    ).one()['value']
+
+
+def get_forms_in_date_range(domain, start, end):
+    XFormInstance.view(
+        "couchforms/all_submissions_by_domain",
+        startkey=[domain, "by_date", start.isoformat()],
+        endkey=[domain, "by_date", end.isoformat(), {}],
+        include_docs=True,
+        reduce=False
+    ).all()
```

Comment:
My thinking was more around performance than anything else.

On 5 May 2015 at 17:37, Daniel Roberts <notifications@github.com> wrote:

> In corehq/ex-submodules/couchforms/dbaccessors.py
> <https://github.com/dimagi/commcare-hq/pull/6524#discussion_r29680806>:
>
> > +    return XFormInstance.view(
> > +        "couchforms/all_submissions_by_domain",
> > +        startkey=startkey,
> > +        endkey=endkey,
> > +        reduce=True,
> > +    ).one()['value']
> > +
> > +
> > +def get_forms_in_date_range(domain, start, end):
> > +    XFormInstance.view(
> > +        "couchforms/all_submissions_by_domain",
> > +        startkey=[domain, "by_date", start.isoformat()],
> > +        endkey=[domain, "by_date", end.isoformat(), {}],
> > +        include_docs=True,
> > +        reduce=False
> > +    ).all()
>
> Thanks: #6566 <https://github.com/dimagi/commcare-hq/pull/6566>. I didn't
> bother with the limit thing because I was trying not to add features that
> weren't used, though you could make a case for it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6524/files#r29680806>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa